x-pack/legacy/plugins/apm/server/routes/ui_filters.ts

```diff
@@ -38,4 +51,143 @@ export function initUIFiltersApi(core: InternalCoreSetup) {
       return getEnvironments(setup, serviceName).catch(defaultErrorHandler);
     }
   });
+
+  const createLocalFiltersEndpoint = ({
+    name,
+    getProjection,
+    validators
+  }: {
+    name: PROJECTION;
+    getProjection: ({
+      setup,
+      query
+    }: {
+      setup: Setup;
+      query: Record<string, any>;
+    }) => Projection;
+    validators?: Record<string, Schema>;
+  }) => {
+    server.route({
+      method: 'GET',
+      path: `/api/apm/ui_filters/local_filters/${name}`,
+      options: {
+        validate: {
+          query: withDefaultValidators({
+            filterNames: Joi.array()
+              .items(localUIFilterNames)
+              .required(),
+            ...validators
+          })
+        },
+        tags: ['access:apm']
+      },
+      handler: async req => {
+        const setup = await setupRequest(req);
+
+        const { uiFilters, filterNames } = (req.query as unknown) as {
+          uiFilters: string;
+          filterNames: LocalUIFilterName[];
+        };
+
+        const parsedUiFilters = JSON.parse(uiFilters);
+
+        const projection = getProjection({
+          query: req.query as Record<string, any>,
+          setup: {
+            ...setup,
+            uiFiltersES: await getUiFiltersES(
+              req.server,
+              omit(parsedUiFilters, filterNames)
+            )
+          }
+        });
+
+        return getLocalUIFilters({
+          server: req.server,
+          projection,
+          setup,
+          uiFilters: parsedUiFilters,
+          localFilterNames: filterNames
+        }).catch(defaultErrorHandler);
+      }
+    });
+  };
+
+  createLocalFiltersEndpoint({
+    name: PROJECTION.SERVICES,
+    getProjection: ({ setup }) => {
+      return getServicesProjection({ setup });
+    }
+  });
```

Comment:
This is going to be a lot harder to type since endpoints are dynamically created. What about extracting the duplicated parts like `handler` and `validators` but still having static endpoints?

Reply:
Eg.
```ts
  server.route({
    method: 'GET',
    path: `/api/apm/...`,
    options: {
      validate: { query: uiFiltersValidators },
      tags: ['access:apm']
    },
    handler: async req => {
      getUIFiltersHandler(req, ...)
    }
  });
```

Reply:
@sqren Ok with you if we explore that in your PR after this is merged?

Reply:
Sure. 
But would be great if they can be converted to static endpoints soon-ish so it can unblock https://github.com/elastic/kibana/pull/42252

Reply:
@sqren yeah, I would be happy to add a commit to your PR once this one lands.